// @nomicfoundation/slang/index.js
"use strict";
// This file is generated automatically by infrastructure scripts. Please don't edit by hand.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.text_index = exports.query = exports.parse_output = exports.parse_error = exports.language = exports.kinds = exports.cursor = exports.cst = exports.ast = void 0;
exports.ast = __importStar(require("./ast"));
exports.cst = __importStar(require("./cst"));
exports.cursor = __importStar(require("./cursor"));
exports.kinds = __importStar(require("./kinds"));
exports.language = __importStar(require("./language"));
exports.parse_error = __importStar(require("./parse_error"));
exports.parse_output = __importStar(require("./parse_output"));
exports.query = __importStar(require("./query"));
exports.text_index = __importStar(require("./text_index"));
